common/flatpak-run.c

```diff
@@ -1339,22 +1329,46 @@ flatpak_run_apply_env_appid (FlatpakBwrap *bwrap,
 }
 
 void
-flatpak_run_apply_env_vars (FlatpakBwrap *bwrap, FlatpakContext *context)
+flatpak_run_apply_env_vars (FlatpakBwrap *bwrap, FlatpakContext *context, const char *app_id)
 {
+  g_autoptr(GHashTable) vars = NULL;
+  int i;
   GHashTableIter iter;
-  gpointer key, value;
+  const char *var;
+  const char *value;
 
-  g_hash_table_iter_init (&iter, context->env_vars);
-  while (g_hash_table_iter_next (&iter, &key, &value))
+  /* hardcode FLATPAK_ID */
+  vars = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_free);
+  g_hash_table_insert (vars, g_strdup ("FLATPAK_ID"), g_strdup (app_id));
+  flatpak_bwrap_set_env (bwrap, "FLATPAK_ID", app_id, TRUE);
+
+  for (i = 0; i < context->env_vars->len; i++)
     {
-      const char *var = key;
-      const char *val = value;
+      var = g_ptr_array_index (context->env_vars, i);
+      value = g_ptr_array_index (context->env_vals, i);
 
-      if (val && val[0] != 0)
-        flatpak_bwrap_set_env (bwrap, var, val, TRUE);
+      /* FLATPAK_ID is readonly */
+      if (g_str_equal (var, "FLATPAK_ID"))
+        continue;
+  
+      if (value && value[0])
+        g_hash_table_insert (vars, g_strdup (var), flatpak_expand_env_vars (value, vars));
```

Comment:
This seems weird. It only expands to the values that are in the context env_var. I expect it to expand against the whole environment that is set so far. I.e. bwrap->envp. 

I was expecting code like:
 flatpak_bwrap_set_env (bwrap, var, flatpak_expand_env_vars (var, bwrap->envp), TRUE);

Reply:
For example, $PATH will be set in the envp before flatpak_run_apply_env_vars, and I expect to be able to extend it.